storage/trace_test.go

```diff
@@ -0,0 +1,302 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/internal/testutil"
+	"cloud.google.com/go/storage/internal"
+	"github.com/google/go-cmp/cmp"
+	"github.com/googleapis/gax-go/v2"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/sdk/trace/tracetest"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/api/option"
+)
+
+type emulatedTraceTest struct {
+	*testing.T
+	name            string
+	resources       resources
+	transportClient *Client
+}
+
+type traceFunc func(ctx context.Context, c *Client, fs *resources) error
+
+func TestTraceStorageTraceStartEndSpan(t *testing.T) {
+	ctx := context.Background()
+	e := tracetest.NewInMemoryExporter()
+	tp := sdktrace.NewTracerProvider(sdktrace.WithSyncer(e))
+	defer tp.Shutdown(ctx)
+	otel.SetTracerProvider(tp)
+
+	// TODO: Remove setting development env var upon launch.
+	t.Setenv("GO_STORAGE_DEV_OTEL_TRACING", "true")
+
+	spanName := "storage.TestTrace.TestStorageTraceStartEndSpan"
+	addAttrs := attribute.String("foo", "bar")
+	spanStartOpts := []trace.SpanStartOption{
+		trace.WithAttributes(addAttrs),
+	}
+	newAttrs := attribute.Int("fakeKey", 800)
+
+	ctx, span := startSpan(ctx, spanName, spanStartOpts...)
+	span.SetAttributes(newAttrs)
+	endSpan(ctx, nil)
+
+	spans := e.GetSpans()
+	if len(spans) != 1 {
+		t.Errorf("expected one span, got %d", len(spans))
+	}
+
+	// Test StartSpanOption and Cloud Trace adoption common attributes are appended.
+	// Test startSpan returns the span and additional attributes can be set.
+	wantSpan := createWantSpanStub(spanName)
+	wantSpan.Attributes = append(wantSpan.Attributes, addAttrs)
+	wantSpan.Attributes = append(wantSpan.Attributes, newAttrs)
+	opts := []cmp.Option{
+		cmp.Comparer(spanAttributesComparer),
+	}
+	for _, span := range spans {
+		if diff := testutil.Diff(span, wantSpan, opts...); diff != "" {
+			t.Errorf("diff: -got, +want:\n%s\n", diff)
+		}
+	}
+	e.Reset()
+}
+
+func TestTraceSpansEmulated(t *testing.T) {
+	checkEmulatorEnvironment(t)
+
+	// Create non-wrapped client to use for setup steps.
+	ctx := context.Background()
+	client, err := NewClient(ctx)
+	if err != nil {
+		t.Fatalf("storage.NewClient: %v", err)
+	}
+
+	skippedTraceMethods := []string{"storage.Bucket.AddNotification", "storage.Bucket.Notifications", "storage.Bucket.DeleteNotification"}
+	for spanName, fn := range traceMethods {
```

Comment:
Yeah I was hoping to do that too. I actually updated to use the `multiTransportTest` harness from the integration_test. 

I couldn't leverage the client_test `transportClientTest` mainly because we're testing trace instrumentation on the transport-agnostic layer, and for testing purposes I also need to configure `NewGRPCClient` and `NewClient` by passing in ClientOption.

Reply:
Ahh got it.